octobot_trading/personal_data/portfolios/portfolio.py

```diff
@@ -261,19 +271,35 @@ def _reset_currency_portfolio_available(self, currency_to_reset, reset_quantity)
         Reset currency portfolio to available
         :param currency_to_reset: the currency to reset
         :param reset_quantity: the quantity to reset
+        :return: boolean to ensure errors forwarding when cythonized
         """
         if reset_quantity is None:
             self._set_currency_portfolio(currency=currency_to_reset,
                                          available=self.portfolio[currency_to_reset][common_constants.PORTFOLIO_TOTAL],
                                          total=self.portfolio[currency_to_reset][common_constants.PORTFOLIO_TOTAL])
-        else:
-            self._update_currency_portfolio(currency=currency_to_reset, available=reset_quantity)
+            return True  # to ensure PortfolioNegativeValueError forwarding when cythonized
+        return self._update_currency_portfolio(currency=currency_to_reset, available=reset_quantity)
 
 
-def _check_available_should_update(order):
+def _should_update_available(order):
     """
     Check if the order has impact on availability
     :param order: The order to check
     :return: True if the order should update available portfolio
     """
     return not order.is_self_managed()
+
+
+def ensure_portfolio_update_validness(currency, origin_quantity, update_quantity):
+    """
+    Ensure that the portfolio final value is not negative.
+    Raise a PortfolioNegativeValueError if the final value is negative
+    :param currency: the currency to update
+    :param origin_quantity: the original currency value
+    :param update_quantity: the update value
+    :return: the updated quantity
+    """
+    if origin_quantity + update_quantity < 0:
+        raise errors.PortfolioNegativeValueError(f"Trying to update {currency} with {update_quantity} "
```

Comment:
I think this can't be raised if we define bint as a return type in header:
`cdef bint _should_update_available(order_class.Order order)`
(test in https://github.com/Drakkar-Software/OctoBot-Benchmarks/tree/master/cython_types#results)

Reply:
Yes right, I forgot, that's why tests are still failing :smile: